### PR TITLE
Add async filtering support to the select-control component

### LINKED ISSUE #35076: __experimentalSelectControl: Add async support
**Description**

The current [experimental select control](https://github.com/woocommerce/woocommerce/tree/trunk/packages/js/components/src/experimental-select-control) forces the consumer of the component to handle all the async logic, this is in theory fine, but will require quite a bit of boiler plate for any async fields.

The Select Control also makes it a bit tricky to notify the user when the dropdown is searching or not.

Here are a couple suggestions that should be explored and/or implemented as part of this issue:
- Allow the `getFilteredItems` callback to support the return of a promise.
- Allow the addition of `useDebounce` for either the `getFilteredItems` function or `onInputChange`, maybe this should be a default of 250ms that can be changed if needed?

See [this PR](https://github.com/woocommerce/woocommerce/pull/34949) and the comments left behind by @mattsherman  for some extra context.

**Acceptance criteria**

- [ ] The `getFilteredItems` prop should support setting it to a Promise
- [ ] The `getFilteredItems` prop should support setting it to a regular function
- [ ] The `getFilteredItems` prop should be debounced, so that it isn't called for every input change
- [ ] A loading/busy indicator should be displayed as the `suffix` when the `getFilteredItems` callback is pending

### PR DESCRIPTION
**Pull request by mdperez86**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/35076

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

1. Go to `Products -> Add New (MVP)` page
2. Run command `cd ./tools/storybook && pnpm run storybook`
3. The experimental select control should perform as expected.
http://localhost:6007/?path=/docs/woocommerce-admin-experimental-selectcontrol--async

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [x] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (joshuatf) [inline packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: Just thinking out loud, I wonder if we should be slightly more opinionated about props returned here.  This might cut down on the amount of boilerplate each consumer needs to write.

We could optionally include:
* a `suffix` (a spinner icon) when we're fetching
* items for each state (loading vs finished)
* a default `getFilteredItems` which just returns all in items fetched

This would house a couple of commonly used state items here, but consumers could always override these if they wanted to.

File: packages/js/components/src/experimental-select-control/stories/index.tsx
Comment (joshuatf) [inline packages/js/components/src/experimental-select-control/stories/index.tsx:162]: If we go with the above recommendation, I'd love to have a simple example in addition to this that doesn't need the `onFilter...` events, but this is a good example to show off its capability.

**Review by joshuatf — COMMENTED** [review]
This looks really great, @mdperez86!  Left a couple optional comments, but I think it may help reduce the effort needed by the consumer to implement.

Would love to get @mattsherman's thoughts on this as well.

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mattsherman) [inline packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: If we set default values for these to no-ops (`() => {}`) then we can remove the checks, which would make the code a bit less noisy.

Though, thinking about it, when not using TS, we really have no guarantee that those are functions if set, so should we check their type before trying to call them? Not sure how defensive we want to be, but it would be nice to have a consistent approach we follow when dealing with function props. What do you think?

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mattsherman) [reply packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: Somewhat related, I was thinking about whether we should passthrough the props here or have the consumer separately set them...

```
<SelectControl< DefaultItemType > { ...selectProps } { ...asyncSelectProps }>
```

But, if we want to be more opinionated as Josh suggests, it could make sense to keep them being passed through as is currently done to allow for that.

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mdperez86) [reply packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: About the default values I'm partially agree with that. My reasoning about that is the following.
```typescript
function main(cb = () => {}) {
  cb();
}
```
if we call the prev function like `main()` or `main(undefinded)` the default value is gonna work but if for some reason we call `main(nullish_variable_not_undefined)` then the default value is not gonna work. Because of that I try to avoid default value attributes as much as possible.

What I described above has a close relation with what you said in your second point, about passing a non callable value, and of course, I'm in the same line as you. So this is defenately something I have to change to:
```typescript
if (typeof cb == 'function') cb();
```
the previous condition prevent the cb function of been called if it's anything else than just a function and also let us to not use default values.

File: packages/js/components/src/experimental-select-control/stories/index.tsx
Comment (mdperez86) [reply packages/js/components/src/experimental-select-control/stories/index.tsx:162]: > If we go with the above recommendation, I'd love to have a simple example in addition to this that doesn't need the onFilter... events, but this is a good example to show off its capability.

Do you mean to a different example using only the `filter` function, ignoring the others because they are optional right?

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mdperez86) [reply packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: > - a suffix (a spinner icon) when we're fetching
> - items for each state (loading vs finished)

I think this can be implemented in a different issue.  Since it requires some design decisions from @jarekmorawski. 

> - a default getFilteredItems which just returns all in items fetched

From my point of view this function should be removed and use the sync version of this hook instead, so it will be a code that is gonna be deleted if I add a default value to getFilteredItems inside the useAsyncFilter hook.

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mattsherman) [inline packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: I'm divided on the naming of `UseAsyncFilterInput` and `UseAsyncFilterOutput`... my initial reaction was that the names didn't feel natural. Though, on the other hand, the use of `Input` and `Output` perfectly describe how a function works. 😄

I wonder if it is just me, or if others had a similar initial reaction? Just sharing my thoughts. No need to change this.

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mattsherman) [inline packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: Should check to ensure that `filter` is a function before calling it.

File: packages/js/components/changelog/add-35076
Comment (mattsherman) [inline packages/js/components/changelog/add-35076]: Instead of `select-control`, consider using `__experimentalSelectControl`.

File: packages/js/components/src/experimental-select-control/index.ts
Comment (mattsherman) [inline packages/js/components/src/experimental-select-control/index.ts:2]: I wonder if `useAsyncFilter` is too generic of an exported name? I just wonder if consumers would end of having to rename it when importing to make things more clear.

File: packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts
Comment (mattsherman) [inline packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts]: Is this needed here?

File: packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts
Comment (mattsherman) [inline packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts:48]: Should we test the specific implementation detail? If the debouncing implementation changed to not use `useDebounce`, the test would fail, even though it might still be debouncing correctly.

File: packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts
Comment (mattsherman) [inline packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts]: I wonder if it would be better in these tests to set `inputValue` to a non-zero-length string. Because, I could see any implementation bug returning `''`, but it would be less likely for it to return something more unique, like `'my filter input'`.

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (joshuatf) [inline packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: Should we add `suffix` in here based on `isFetching` status?

>A loading/busy indicator should be displayed as the suffix when the getFilteredItems callback is pending

I think this is the only criteria left in the original issue that needs to be resolved.

File: packages/js/components/src/experimental-select-control/stories/index.tsx
Comment (joshuatf) [inline packages/js/components/src/experimental-select-control/stories/index.tsx:213]: Thanks for the additional example!

File: packages/js/components/src/experimental-select-control/stories/index.tsx
Comment (joshuatf) [inline packages/js/components/src/experimental-select-control/stories/index.tsx]: My earlier comment was about moving this into the `useAsyncFilter` return.  If we decide to remove this later we can do that, but it might be nice for consumers to not have to know how to do this.

**Review by joshuatf — COMMENTED** [review]
I'm having some trouble getting my Storybook working with this branch.  Not sure if it's my local env or not.

Code-wise this looks to me.  I added a couple comments I'd love to see added in to resolve the original issue, but otherwise this has my approval.

I'll let @mattsherman give final ✅  if he can get Storybook working since I know he has a pending review.

**Review by mattsherman — CHANGES_REQUESTED** [review]
Great iteration on this so far, @mdperez86 !

I went through and reviewed all the code. Left some comments... most are just suggestions/questions.

I can't get the Storybook build to run, so I was unable to test that out. Once any additional changes are made and the Storybook build is fixed, I can re-review and run that.

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mdperez86) [reply packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: Also I don't like to render components inside custom hooks like this but I can add it anyway.
```jsx
suffix: isFetching ? <Spinner /> : <DefaultIcon />,
```

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mdperez86) [reply packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: If you had to use the types outside of the hook how would you call them?
I know those names are very uhmm don't know! But I'm super sure I can't call them `Input` and `Output`, inside the hook it's ok but outside?
Do you have any idea here?

File: packages/js/components/src/experimental-select-control/index.ts
Comment (mdperez86) [reply packages/js/components/src/experimental-select-control/index.ts:2]: Agree, what is your suggestion here?

File: packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts
Comment (mdperez86) [reply packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts:48]: Here I'm not testing the execution/functionality of the `useDebounce` hook, it's just if it was called with the right set of params like the delay time. 
In my opinion we don't need to test external function in unit tests that's why we mocked them. So if we decide to use a diff implementation it's ok that the test fail so we can fix it with the new implementation.
But if you think `useDebounce` should be tested transitively then this is not an unit test and it is more like an integration test. Should I write integration tests instead?

File: packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts
Comment (mdperez86) [reply packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts]: Yes! the hook expects a function that returns a Promise. What do you think is gonna happen if the `filter` function returns `jest.fn()`? We cannot do `jest.fn().then(...)` right?

**Comment by mdperez86** [comment]
Blocked by https://github.com/woocommerce/woocommerce/pull/35867

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mattsherman) [reply packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: That sounds good to me.

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mattsherman) [reply packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: I wonder if we should leave the `suffix` totally up to the consumer? But, I suppose they can always just omit the returned `suffix` if they don't want our auto-handling of it.

File: packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts
Comment (mattsherman) [reply packages/js/components/src/experimental-select-control/hooks/use-async-filter.ts]: Sorry for the confusion... I wasn't suggesting to call them `Input` and `Output`... I was just saying that "input" and "output" makes sense in the names even though initially they felt "off" to me. 

I say we keep these names... they are growing on me.

File: packages/js/components/src/experimental-select-control/index.ts
Comment (mattsherman) [reply packages/js/components/src/experimental-select-control/index.ts:2]: What about `useSelectControlAsyncFilter`?

File: packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts
Comment (mattsherman) [reply packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts]: Oh, of course. 😄

File: packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts
Comment (mattsherman) [reply packages/js/components/src/experimental-select-control/test/use-async-filter.spec.ts:48]: I don't think we should test the functionality of the `useDebounce` hook. But, this expect is really just ensuring that the implementation used `useDebounce`. I guess it's okay... ideally I'd like to test the debouncing. I tried something similar in the tests for `<DateTimePickerControl>`, but I had to skip the test, so perhaps it isn't really easily testable (or, I just don't know how to do it!)...

https://github.com/woocommerce/woocommerce/blob/0a039a1f1315cd4a212957cac2bc29771bbba234/packages/js/components/src/date-time-picker-control/test/index.tsx#L582

So, we could just keep the expect as is for now... it would be clear why the test fails if we were to change the implementation to not use `useDebounce`, so not a big deal.

**Review by mattsherman — APPROVED** [review]
I tested in Storybook and things behave as expected.

Nice work on this -- I think this is good to be merged.

**Pull request merged**
